main.tf

```diff
@@ -1,13 +1,29 @@
-#
-# Developer tips:
-#   - Below code should be replaced with the code for the root level module
-#
+data "ibm_is_subnet" "selected" {
+  count      = length(var.subnet_ids) # Use count instead of for_each
+  identifier = var.subnet_ids[count.index]
+}
+
+locals {
+  baremetal_servers = { for idx in range(var.server_count) : idx => {
+    prefix    = var.server_count == 1 ? var.prefix : "${var.prefix}-${idx}"
+    subnet_id = var.subnet_ids[idx % length(var.subnet_ids)]
+    zone      = data.ibm_is_subnet.selected[idx % length(var.subnet_ids)].zone
+  } }
+}
+
+module "baremetal" {
+  source   = "./modules/baremetal"
+  for_each = local.baremetal_servers
```

Comment:
looks like the index for this will be an integer based on server_count.

The challenge of doing these kinds of dynamic modules is to make sure that too many resources are not destroyed when input values change.

For example, I initially give server_count = 5 and a list of 3 subnets. What will happen if I remove the FIRST subnet in the list? Will it destroy every single instance? Or will it only destroy instances that were in that first subnet, and add extra servers to the remaining subnets based on server_count?

You will need to test what happens when either of those inputs change, and make sure that instances are not unintentionally destroyed because of the way the mapping is done on the terraform state side.

Reply:
Ran a test starting with count = 2 and 2 subnets. If I removed first subnet and set count=1 it resulted in destruction of ALL existing servers, and created a new one.